### PR TITLE
Add refactoring topic to Building an API series

### PR DESCRIPTION
**Pull request by kelsey-steven-ada**
NOTE: This PR does not contain changes to Building an API parts 7, 8, & 9 to update the numbering sequences and branch names since this lesson will be the new pt. 7. Those updates have been made in a separate PR so this one can be focused on just the new content ([Link to PR for those changes](https://github.com/Ada-Developers-Academy/core-unit-2/pull/127)).

Any feedback is appreciated, I was working off the notes under [Unit 2: Refactoring in this curriculum doc](https://docs.google.com/document/d/1cNDAgiSHljfICUv5fjh5nh7dgu9uh12BrrcmzJGJlsw/edit?usp=sharing).
I don't know if I will have time for more thorough "Check for Understanding" questions, so I put in the default "Tell us about something you took away" question for now. 

This topic has 4 lessons:
| Lesson | Content |
|--|--|
| What is Refactoring? | General information about how and why we refactor |
| Extracting Helper Functions Pt. 1 | Create a `to_dict` instance method in `book.py` |
| Extracting Helper Functions Pt. 2 | Create a `from_dict` class method function in `book.py` |
| Refactoring `validate_book` | Refactors `validate_book` into a more generic function `validate_model` |

### CODE REVIEW

**Review by kyra-at-ada — COMMENTED** [review]
This is just a content/language review as I'm having issues with running the tests and previewing the entire lesson (I was able to preview individual lessons). I'll continue working on getting that resolved today for a full review.

This lesson looks great! It would be okay to publish even without accepting any changes. I really liked the way you structured the lessons, motivating each refactor before working through the implementation.

All of the suggestions are very minor. Most of them focus on eliminating passive voice and ambiguous object references in sentences such as 'it' to increase clarity. 

I might also consider taking a second read through the lessons and changing language that uses 'if' to a more direct suggestion where possible. For example changing language currently phrased as, 'If we run the tests one last time, we should see all the tests passing' to 'Running the tests  one last time, check that all the tests now pass." This will create a stronger implication that running the tests is a direction students should copy rather than a suggestion or optional step. 

Really great work!

File: api-7-refactoring/extracting-helper-functions-from-dict.md [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+| Class Method | A function that is part of the class definition that only has access to class-level variables. It must be prefixed with the decorator `@classmethod` and takes `cls` as its first parameter. They can be called by referencing the class name followed by the function name (Ex.  `<Class_Name>.<function_name>()`) | Class Function | "We created a class method that returns a new instance with all the attributes set to 0" |
+
+## Why refactor what isn't repeated?
+
+We know that we have code in the `routes.py` file that creates an instance of a `Book` model from a dictionary, and we plan to refactor it to the `Book` class, but _why_ do we want to do that?
+
+So far we've refactored to reduce repeated code, but there are so many other reasons we might want to refactor! A common reason is improving our code organization. We often want functions that operate directly on a particular class to be bundled with that class's code. We save time by having those functions ship with the class and be easy to find and navigate.
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-from-dict.md:54]: I have to keep looking this up because it always looks wrong to me, but singular possessive of "class" is "class's". *Shakes fist at the English language*

File: api-7-refactoring/extracting-helper-functions-from-dict.md [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+
+We know that we have code in the `routes.py` file that creates an instance of a `Book` model from a dictionary, and we plan to refactor it to the `Book` class, but _why_ do we want to do that?
+
+So far we've refactored to reduce repeated code, but there are so many other reasons we might want to refactor! A common reason is improving our code organization. We often want functions that operate directly on a particular class to be bundled with that class's code. We save time by having those functions ship with the class and be easy to find and navigate. 
+
+There are many kinds of functions we package with classes, but let's focus on initializers for a moment. All classes have some kind of initializer, whether it's the default that the language provides or one we write ourselves. "Convenience Initializers" are functions other than the default initializer that let us perform some extra logic or set up defaults to create an instance of a class. Tying it back to `Hello Books`, if we know that we'll frequently have a dictionary and want to create a `Book` from its contents, we might want a convenience initializer that ships with the `Book` class to handle the task.
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-from-dict.md:56]: Thanks for the heads up, I'll keep that in mind for the future ^_^

[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-from-dict.md:56]: Great point... What about "bundled with"? 

"... we might want a convenience initializer bundled with the `Book` class to handle the task."

File: api-7-refactoring/extracting-helper-functions-from-dict.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (kyra-at-ada) [reply api-7-refactoring/extracting-helper-functions-from-dict.md:54]: Ah yes this is indeed singular not plural. Nice catch of my attempt!

File: api-7-refactoring/extracting-helper-functions-from-dict.md [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+```
+
+</details>
+</br>
+
+If we run our test suite and see only these new tests are failing, we can start implementing!
+
+### Write the `from_dict` function
+
+Our next step would be to write out our pseudocode. We'll leave that as an exercise for you to try out and move on to writing our function definition.
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-from-dict.md:267]: I like that, this would be a nice place for extra engagement.

File: api-7-refactoring/extracting-helper-functions-to-dict.md [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+
+- A `hello_books_development` database
+- A `book` table defined
+- A `Book` model defined
+- Endpoints defined for these RESTful routes:
+- `GET` to `/books`
+- `POST` to `/books`
+- `GET` to `/books/<book_id>`
+- `PUT` to `/books/<book_id>`
+- `DELETE` to `/books/<book_id>`
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-to-dict.md:33]: I also would prefer a sublist here. This content is from the first lesson in topic 7 of Building an Api, so there'll be a slight change in the formatting between lessons, but I can't imagine that being an issue here.

File: api-7-refactoring/extracting-helper-functions-to-dict.md [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+
+The responses are a little different, but in both cases we need to convert a `Book` model to a dictionary to create our response. We can see this code repeated in both functions, so let's move forward with refactoring it to a helper function!
+
+### Identifying Dependencies
+
+We've identified the code we want to refactor, but do we have the safety net we need to refactor confidently? To know that, we need to first identify our dependencies, then ensure we have strong test coverage around those scenarios.
+
+To identify dependencies, we want to look at the code base and ask "Where is the code I'm refactoring being called?". We should keep track of all the locations we find. If we're refactoring code inside a class, we'd also want to ask "Does any other code extend or inherit this class?".
+
+In our case, we can search the project and feel confident that our two read functions, `read_all_books` and `read_one_book`, are the only dependencies.
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-to-dict.md:90]: Great point, I'll add something in

File: api-7-refactoring/refactor-validate-book.md [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+
+Taking a look over `routes.py`, we can see that all of the routes that read or alter a model by `id` use the `validate_book` function. This makes sense, we're reusing code to fetch a single `Book` from the database, so that's all good right? Here's where we want to think about our upcoming models. 
+
+When we add a new model, say `Author`, and create routes for it, what will we need to do for routes that read, update or delete an `Author` by `id`? If we don't have a function flexible enough to work for any model, then we may need a custom validate function for every model we add. In a very small project, this may not be cumbersome, but as a project grows it can become a pain point, especially if there is ever a need to update the pattern for fetching a model.
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/refactor-validate-book.md:61]: I like this reorganization, much clearer what is being asked and why 😄

File: api-7-refactoring/refactor-validate-book.md [hunk 7]
```diff
@@ -0,0 +1,421 @@
...
+
+```
+Failed: [undefined]AssertionError: assert {'message': '... 3 not found'} == {'message': '... 3 not found'}
+  Differing items:
+  {'message': 'Book 3 not found'} != {'message': 'book 3 not found'}
+```
+
+Taking this line-by-line, we can see that we have an `AssertionError` around the `message` returned. The test was expecting `{'message': 'book 3 not found'}`, but it got a response with different capitalization: `{'message': 'Book 3 not found'}`. Our change `cls.__name__` gives us the exact name of our class, which is capitalized, so the string is different from what our tests expect! 
+
+We could change our `validate_book` code to lowercase the result of `cls.__name__` so our tests pass again, but in this case we're going to choose to update our tests and keep the new capitalization. Once we've gone through and updated the expected messages for our failing tests, we should see everything passing again.
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/refactor-validate-book.md:332]: It could be worth having if folks might fixate on it. My only reasoning was that it's the start of a sentence and reflects the class name's actual capitalization. I'm thinking about it as additional info that isn't important to the overall process of refactoring, would that make sense to go in an information block right after?

File: api-7-refactoring/what-is-refactoring.md [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+
+At this point we should start a cycle of making a single small improvement, then running our test suite to ensure our functionality hasn’t changed. If we make our changes and our tests are failing, we should practice our debugging skills, making use of the debugger and tests to help pinpoint what’s changed.
+
+As we go, we may find that we need to add more tests, or pivot our approach slightly. All of this is normal, and as long as we have a strong test suite, we can continue to iterate and have peace of mind that our code is still behaving as expected. 
+
+### !callout-info
+
+## Tests give us the security to refactor with confidence!
+
+Without tests, we could still make changes, but we couldn’t use automation to help us check if there were side affects. We would have to take the time to manually test each scenario every time we made a change if we wanted to know that we had not broken something.
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/what-is-refactoring.md:69]: Hmm, I'm not sure if reiteration on unit vs manual testing is helpful here... The Refactoring topic will come pretty late in Unit 2 and students will have had several projects with both unit test suites and play tester files for manual testing by then. I'll try to get opinions from a couple other instructors based on what they're seeing with students.

File: api-7-refactoring/refactor-validate-book.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kyra-at-ada) [reply api-7-refactoring/refactor-validate-book.md:334]: Maybe just specify that either choice is valid? I think an info block will just draw more attention to it. I just think the language here implies our choice might be slightly better

File: api-7-refactoring/refactor-validate-book.md [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
```
Comment (kelsey-steven-ada) [inline api-7-refactoring/refactor-validate-book.md:334]:
```suggestion
We could change our `validate_book` code to lowercase the result of `cls.__name__` so our tests pass again, but in this case we're going to choose to update our tests and keep the new capitalization; either choice is valid for `hello_books`. In the industry, and generally when working with other people, we would consult with others invested in the project before changing user-facing messages. Once we've gone through and updated the expected messages for our failing tests, we should see everything passing again.
```

File: api-7-refactoring/refactor-validate-book.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/refactor-validate-book.md]: I added some text clarifying that either choice is valid for `hello_books` and a sentence to the effect of "generally we'd be consulting with people before changing user messages".

File: api-7-refactoring/extracting-helper-functions-to-dict.md [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+        books_response.append(book.to_dict())
+    return jsonify(books_response)
+```
+
+</details>
+</br>
+
+Once we've completed changes for `read_all_books`, we'll follow the same steps for `read_one_book`:
+1. Delete the old dictionary creation code
+2. Try running tests and see them fail
```
Comment (kelsey-steven-ada) [inline api-7-refactoring/extracting-helper-functions-to-dict.md:332]:
```suggestion
2. Run unit tests and see them fail
```

File: api-7-refactoring/extracting-helper-functions-to-dict.md [hunk 11]
```diff
@@ -0,0 +1,370 @@
...
+   <summary>When you're done brainstorming, expand this section to review our cases and new test code. </summary>
+
+We want to add tests that check:
+- When we have records, `read_all_books` returns a list containing a dictionary representing each `Book`
+- When we have records and a `title` query in the request arguments, `read_all_books` returns a list containing only the `Book`s which match the query
+- When we call `read_one_book` with a numeric ID that doesn't have a record, we get the expected error message
+- When we call `read_one_book` with a non-numeric ID, we get the expected error message
+
+```python
+def test_get_all_books_with_two_records(client, two_saved_books):
```
Comment (kelsey-steven-ada) [inline api-7-refactoring/extracting-helper-functions-to-dict.md:110]:
```suggestion
# When we have records, `read_all_books` returns a list containing a dictionary representing each `Book`
def test_get_all_books_with_two_records(client, two_saved_books):
```

File: api-7-refactoring/extracting-helper-functions-to-dict.md [hunk 12]
```diff
@@ -0,0 +1,370 @@
...
+        "title": "Ocean Book",
+        "description": "watr 4evr"
+    }
+    assert response_body[1] == {
+        "id": 2,
+        "title": "Mountain Book",
+        "description": "i luv 2 climb rocks"
+    }
+
+def test_get_all_books_with_title_query_matching_none(client, two_saved_books):
```
Comment (kelsey-steven-ada) [inline api-7-refactoring/extracting-helper-functions-to-dict.md:129]:
```suggestion
# When we have records and a `title` query in the request arguments, `read_all_books` returns a list containing only the `Book`s which match the query
def test_get_all_books_with_title_query_matching_none(client, two_saved_books):
```

File: api-7-refactoring/extracting-helper-functions-to-dict.md [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+    # Act
+    data = {'title': 'Desert Book'}
+    response = client.get("/books", query_string = data)
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 200
+    assert response_body == []
+
+def test_get_all_books_with_title_query_matching_one(client, two_saved_books):
```
Comment (kelsey-steven-ada) [inline api-7-refactoring/extracting-helper-functions-to-dict.md:139]:
```suggestion
# When we have records and a `title` query in the request arguments, `read_all_books` returns a list containing only the `Book`s which match the query
def test_get_all_books_with_title_query_matching_one(client, two_saved_books):
```

File: api-7-refactoring/extracting-helper-functions-to-dict.md [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+    # Assert
+    assert response.status_code == 200
+    assert len(response_body) == 1
+    assert response_body[0] == {
+        "id": 1,
+        "title": "Ocean Book",
+        "description": "watr 4evr"
+    }
+
+def test_get_one_book_id_not_found(client, two_saved_books):
```
Comment (kelsey-steven-ada) [inline api-7-refactoring/extracting-helper-functions-to-dict.md:154]:
```suggestion
# When we call `read_one_book` with a numeric ID that doesn't have a record, we get the expected error message
def test_get_one_book_id_not_found(client, two_saved_books):
```

File: api-7-refactoring/extracting-helper-functions-to-dict.md [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+    # Act
+    response = client.get("/books/3")
+    response_body = response.get_json()
+
+    # Assert
+    assert response.status_code == 404
+    assert response_body == {"message":"book 3 not found"}
+
+def test_get_one_book_id_invalid(client, two_saved_books):
```
Comment (kelsey-steven-ada) [inline api-7-refactoring/extracting-helper-functions-to-dict.md:163]:
```suggestion
# When we call `read_one_book` with a non-numeric ID, we get the expected error message
def test_get_one_book_id_invalid(client, two_saved_books):
```

**Review by kyra-at-ada — COMMENTED** [review]
Biggest note is that branch `07a-to-dict-refactor` doesn't include the changes from the lesson. 

A couple of other small notes, but overall the code and branches look good!

File: api-7-refactoring/extracting-helper-functions-from-dict.md [hunk 16]
```diff
@@ -0,0 +1,338 @@
...
+- Create a class method named `from_dict` in `book.py` 
+- Refactor the code for creating a `Book` model in the `create_book` route to use `from_dict`
+
+To do this we will follow the same steps of planning then executing a refactor as outlined in the previous lesson. 
+
+## Branches
+
+| Starting Branch | Ending Branch|
+|--|--|
+|`07a-to-dict-refactor` |`07b-from-dict-refactor`|
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-from-dict.md:15]: For clarity, which lesson is 07a not retaining changes for? Before this lesson, part 6 of building an API adds the 3 initial tests which are present on 07a. Am I misunderstanding what lesson you're referring to?

[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-from-dict.md:15]: In case you were looking at them, I haven't updated any of the original lesson 7 (One-to-many) branch names or content since I didn't want to alter them if for some reason this content wasn't merged.

File: api-7-refactoring/extracting-helper-functions-to-dict.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/extracting-helper-functions-to-dict.md]: Added a short answer block with an explanation that shows which dependencies we identified.

File: api-7-refactoring/refactor-validate-book.md [hunk 17]
```diff
@@ -0,0 +1,421 @@
...
+- usage in `update_book`
+- usage in `delete_book`
+
+We can see our dependencies are the functions `read_one_book`, `update_book`, and `delete_book`, so let's see what our tests look like for those functions.
+
+### Check for Tests
+
+We previously wrote tests for `read_one_book`, so we can feel good about that function's coverage, but we don't have any tests for the functions `update_book` or `delete_book`. Furthermore, we don't have tests for `validate_book` to confirm its behavior as we make modifications. 
+
+Let's pause and think through nominal and edge cases for the functions `update_book`, `delete_book`, and `validate_book`. We'll add tests for these functions to `test_routes.py`. Give writing the tests a try, then check out the tests we included below.
```
[the remark this answers is not in the document]
Comment (kelsey-steven-ada) [reply api-7-refactoring/refactor-validate-book.md]: Added a short answer challenge to brainstorm missing test cases to get folks involved and have some division between identifying tests cases and writing them.

**Comment by kelsey-steven-ada** [comment]
@kyra-at-ada is there a curriculum process for new topics once the majority of PR feedback has been addressed? I'm not sure when it's okay for me to merge this work in and wanted to ensure there wasn't something outstanding that should be addressed before the PR can be approved. 

The remaining tasks I'm aware of for the Refactoring topic are: 
- merge this PR, 
- merge the lesson renumbering PR, 
- Update the `hello_books` lesson branches for the current lessons 7-9 (need to be renumbered to 8-10 and all need the refactoring code changes and new tests added).

**Comment by kyra-at-ada** [comment]
> 

Let's go ahead and move folks onto a c18 branch for core-unit-2 curriculum before we merge this into main. That way we can verify correctness in the core sandbox and make sure instructors are okay with adding this content in for C18 before revealing it to students.

let's hold off on updating the lesson branches until that point.
